### DISCUSSION TITLE
General confusion around `--buildtype=plain`

### DISCUSSION THREAD

**Discussion by mtribiere** [mesonbuild/meson#11455] [Q&A] [answered]
I'm trying to cross-compile some C code with a GCC-based compiler for another architecture. As I want complete control over my flags. I assumed that `--buildtype=plain` would get rid of all defaults flags like `-D_FILE_OFFSET_BITS=64`. But it doesn't.

So my question is : Is there any way to remove **all** flags leaving the user specifing all arguments by himself ?

My command :
`$> python ..\meson-1.0.1\meson.py setup --cross-file=..\meson-1.0.1\cross\myCrossCompileProfile.txt --buildtype=custom builddir`

**Comment by eli-schwartz** [comment, marked as answer]
The purpose of buildtype=plain is to remove all flags that would have been set by a buildtype, not to remove all flags, no matter what. There are other flags we always set, such as `/nologo /Zc:__cplusplus` for the Microsoft compiler (the latter flag just instructs MSVC to actually be truthful in the __cplusplus define instead of lying regardless of std and pretending to be c++98 for compatibility with extremely old, extremely broken code).

We also emit the standard depfile arguments for GCC, and you cannot get rid of those either, they're vital for buildsystem tracking.

What's wrong with -D_FILE_OFFSET_BITS=64?

**Reply by mtribiere** [reply]
There is nothing wrong with `-D_FILE_OFFSET_BITS=64` in particular, it's just the fact that for some industrial applications, you sometime wants to have complete control over your flags. And therefore flags that are "impossible to remove no matter what" makes the use of Meson in such cases unsuitable. 

Maybe Meson's philosophy of automating as much as possible is incompatible with that thought.

**Reply by xclaesse** [reply]
Quoting Meson source code:
```python
# If this breaks your code, fix it! It's been 20+ years!
return ['-D_FILE_OFFSET_BITS=64']
```

**Reply by mtribiere** [reply]
Yeah saw that, I wish

**Reply by mtribiere** [reply]
Out of curiosity, do you now where to remove the code that includes `-I. -I..` by default ?

**Reply by eli-schwartz** [reply]
This is probably just https://github.com/mesonbuild/meson/issues/3049 tbh.

Following this discussion, you can pass `-U_FILE_OFFSET_BITS` in a cross file c_args to reset the value.

There may be some way to generally solve this -- platform detection probes sounds like a good idea -- but what I *can* say is that the buildtype profile is the wrong place to handle that, because it's about release vs. debug builds and the options which tie into that such as debuginfo, optimization, CRT selection, disabling asserts, and so on.

**Reply by xclaesse** [reply]
> Out of curiosity, do you now where to remove the code that includes `-I. -I..` by default ?

You can add to your targets: `implicit_include_directories: false`.
